libobs/media-io/media-remux.c

```diff
@@ -149,13 +149,20 @@ static inline bool init_output(media_remux_job_t job, const char *out_filename)
 		av_dict_copy(&out_stream->metadata, in_stream->metadata, 0);
 
 #if LIBAVCODEC_VERSION_INT >= AV_VERSION_INT(57, 48, 101)
-		if (in_stream->codecpar->codec_tag != 0) {
-			out_stream->codecpar->codec_tag =
-				in_stream->codecpar->codec_tag;
-		} else if (in_stream->codecpar->codec_id == AV_CODEC_ID_HEVC) {
+		if (in_stream->codecpar->codec_tag == 0 &&
+		    in_stream->codecpar->codec_id == AV_CODEC_ID_HEVC) {
 			// Tag HEVC files with industry standard HVC1 tag for wider device compatibility
 			out_stream->codecpar->codec_tag =
 				MKTAG('h', 'v', 'c', '1');
+		} else if (!job->ofmt_ctx->oformat->codec_tag ||
+			   av_codec_get_id(job->ofmt_ctx->oformat->codec_tag,
+					   in_stream->codecpar->codec_tag) ==
+				   out_stream->codecpar->codec_id ||
+			   av_codec_get_tag(job->ofmt_ctx->oformat->codec_tag,
+					    in_stream->codecpar->codec_id) <=
+				   0) {
```

Comment:
av_codec_get_tag returns an unsigned integer. 0 if no tag found, otherwise non-zero. Are you sure you meant to check for 0? or did you mean to check for non-zero?

Reply:
I noticed this problem when I ported the code from `segment.c`, ffmpeg Repo.
I think it check for 0, not for non-zero.

should I repush this PR to modify this.

Reply:
I'm currently trying to logically parse what's happening here, please correct me if I got this wrong - based on the change the following should happen:

* When the input stream has no codec tag but the codec id is HEVC, then set the codec_tag to `hvc1`.
* When the output format has no codec tag set, OR
  *  When the input stream has a codec tag, that codec tag has a supported codec id in the output format, and that codec id is equal to the codec id of the output stream, OR
  * When the output format has a supported codec tag given the codec id of the input stream, THEN
  * Set the codec tag of the output stream to the codec tag of the input stream
*  Otherwise set the codec tag to `0`

Shouldn't the compatibility check happen in every case then? Because right now we set the codec tag to `hvc1` regardless of whether the output format supports it (iirc `mkv` might not support it) and with your change only an input stream with an existing tag gets these checks.

Reply:
Also, maybe I'm missing something, but if an output format supports `hev1` but not `hvc1` (for whatever reason), wouldn't the third check yield `true`? You pass in `AV_CODEC_ID_HEVC` to `av_codec_get_tag` and might get `hev1`, wouldn't setting it to `hvc1` break the file then?

Also FFmpeg's code checks for "less then or equal to zero", which means it requires a mismatch ("no tags for codec id found in output format"), which confuses me a bit.

Reply:
Thanks for your analysis, I found some flaws in this PR.

Back to the version before this PR:
Has Tag, directly equal assignment.
Not Has Tag, But `codec_id == HEVC`, assign `tag` to `hvc1`(for wider compatibility)
Else assign `tag` to `0`(ffmpeg will auto select in init_muxer)

Normal version(obs 28):
assign `codec` to `0` in any case.(ffmpeg will auto select)

https://github.com/FFmpeg/FFmpeg/blob/51efa68ec0b4f42b5b124b8987fb68f60a929c4f/libavformat/mux.c#L360-L381


So solution 1:
For wider compatibility:
1. Check OutputTags exist
2. CodecID equal HEVC
3. Check Codec support HVC1

then assign `out stream tag` to `HVC1`(for apple support? )

otherwise, set `0` (ffmpeg auto select)

Or solution 2:
Check and Select outstream tag as possible in OBS remux.

I think solution 1 is much better.
What do you think? Thx.
Look forward to your reply.

Reply:
@PatTheMav 
Excuse me, is this code block for `respect input tags`?
https://github.com/obsproject/obs-studio/commit/bbb3dc00590a572cf49f97db3397221f0828f9ad#R152-R154
```c++
if (in_stream->codecpar->codec_tag != 0) {
			out_stream->codecpar->codec_tag =
				in_stream->codecpar->codec_tag;
		}
```
So need remain this code block and add compatibility check if I understand right?

Reply:
Yeah that one just naively retains the tag - it might have been there before my change even (haven't checked yet), but from the sound of it it could also break the output file if the selected output format does not support the tag.

Given that most people convert straight to `mp4` (and don't change the suffix in the remux window) it might not have occurred that often.

Reply:
> Thanks for your analysis, I found some flaws in this PR.
> 
> Back to the version before this PR: Has Tag, directly equal assignment. Not Has Tag, But `codec_id == HEVC`, assign `tag` to `hvc1`(for wider compatibility) Else assign `tag` to `0`(ffmpeg will auto select in init_muxer)
> 
> Normal version(obs 28): assign `codec` to `0` in any case.(ffmpeg will auto select) _avformat_write_header -> init_output.. -> init_muxer_ https://github.com/FFmpeg/FFmpeg/blob/51efa68ec0b4f42b5b124b8987fb68f60a929c4f/libavformat/mux.c#L360-L381
> 
> So solution 1: For wider compatibility:
> 
> 1. Check OutputTags exist
> 2. CodecID equal HEVC
> 3. Check Codec support HVC1
> 
> then assign `out stream tag` to `HVC1`(for apple support? )
> 
> otherwise, set `0` (ffmpeg auto select)
> 
> Or solution 2: Check and Select outstream tag as possible in OBS remux.
> 
> I think solution 1 is much better. What do you think? Thx. Look forward to your reply.

@PatTheMav If you perfer to `solution 2 : Check and Select outstream tag as possible in OBS remux.`, I will update soon. Thanks for your guidance in advance.